rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,593 @@
+represented by filenames, and a SplitProvider produces `["file1", "file6",
+"file11"]`, an iterator initialized by that `SplitProvider` should process those
+three files only.
+
+```cpp
+class SplitProvider {
+ public:
+  virtual Status GetNext(std::unique_ptr<Split>* split,
+                         bool* end_of_splits) = 0;
+};
+```
+
+When processing datasets, tf.data service workers will use `SplitProvider`s
+which provide splits by querying the tf.data service master for which splits to
+process. A few splits will be prefetched to hide the latency of needing to
+request a new split from the master.
+
+#### Supported Datasets
+
+Not all dataset sources and transformations are easily splittable. For example,
+`take`, `skip`, and `scan` require a global view of the dataset to produce
+correct results. Datasets which require multiple input datasets such as `zip`
+are also difficult to support, since we don't have a good way of aligning the
+splits of multiple input datasets. Users who rely on these unsupported datasets
+will need to move those datasets to come after the distributed part of their
+pipeline.
+
+Initially, we will support splitting for the following dataset sources and
+transformations:
+
+*   `batch`, `CsvDataset`, `dense_to_sparse_batch`, `filter`,
+    `FixedLengthRecordDataset`, `flat_map`, `from_tensor_slices`,
+    `group_by_window`, `ignore_errors`, `interleave`, `list_files`, `map`,
+    `range`, `repeat`, `padded_batch`, `prefetch`, `shuffle`, `SSTableDataset`,
+    `TextLineDataset`, `TFRecordDataset`, `unbatch`, `window`.
+
+### Master and worker services
+
+This section discusses the design for the master and worker services. These
+services are used by the Python API to provide distributed dataset processing,
+and these services use the splitting API as a part of their implementation.
+
+#### Master API
+
+The master is responsible for registering datasets, generating and tracking
+iteration and worker ids, and generating dataset splits for processing on
+workers.
+
+Below is a sketch of the Master API. This API is not public and is subject to
+change.
+
+```cpp
+/// ---- Methods called by consumers ----
+
+// Registers a dataset and returns an id for the dataset. If the dataset is
+// already registered, its dataset id is returned.
+int GetOrRegisterDataset(GraphDef dataset);
+
+// Creates and returns `num_consumers` iterator ids which partition the
+// specified dataset. This also creates an internal `iteration_id` used to
+// track the overall dataset iteration. `num_tasks` defines how many tasks to
+// create. If `num_tasks` is -1, it is up to the master to determine how many
+// tasks to create.
+list<int> CreateIterators(int dataset_id, int num_consumers,
+                          int num_tasks);
+
+// Returns the list of tasks processing data for `iterator_id`. Consumers query
+// this to find which worker addresses to read data from.
+list<TaskInfo> GetWorkersForiterator(int iterator_id);
```

Comment:
Could you please give more details about how to decide which workers the consumer should read data from? Could we enable the end-users to customize the strategy?

nit: `GetWorkersForiterator` -> `GetWorkersForIterator`.

Reply:
Initially, the master will try to schedule one task per worker. Ideally we can make the scheduling smart enough that users don't need to worry about it, but if we can't achieve that, we can consider letting the user specify a scheduling strategy.